src/Moryx.Products.Management/Implementation/Storage/ProductStorage.cs

```diff
@@ -574,17 +576,20 @@ where links.All(l => l.Id != link.Id)
                             linkEntity = typeEntity.Parts.First(p => p.Id == link.Id);
                         }
                         linkStrategy.SavePartLink(link, linkEntity);
-                        linkEntity.Child = GetPartEntity(uow, link);
+                        linkEntity.Child = GetPartEntity(ref saverContext, link);
                     }
                 }
             }
 
             return typeEntity;
         }
 
-        private ProductTypeEntity GetPartEntity(IUnitOfWork uow, IProductPartLink link)
+        private ProductTypeEntity GetPartEntity(ref ProductPartsSaverContext saverContext, IProductPartLink link)
         {
-            return link.Product.Id == 0 ? SaveProduct(uow, link.Product) : uow.GetEntity<ProductTypeEntity>(link.Product);
+            var product = saverContext.EntityCache.ContainsKey((ProductIdentity)link.Product.Identity)?
+                saverContext.EntityCache[(ProductIdentity)link.Product.Identity] :
+                link.Product.Id == 0 ? SaveProduct(ref saverContext, link.Product) : saverContext.UnitOfWork.GetEntity<ProductTypeEntity>(link.Product);
+            return product;
```

Comment:
Just one annotation from me: This line makes too much things: Load from cache if not existing, check cache, save if id is 0 and save if is not 0.  

@Toxantron This maybe fixes problems with a single product type, but what about a complete list of types to save?

Reply:
Since each instance is saved individually this works for both. We could speed things up with a `SaveAll` overload, but in any case the context class and entity cache will work.
Right now the issue was also caused by a single imported instance, that references the same part multiple times. Each instance after could load the type from the database by id.